### PR TITLE
Port over Dafny std library

### PR DESCRIPTION
**Pull request by elaustell**
Having a vast standard library can help prevent reinventing the wheel when it comes to proving behavior about built-in types. This PR will add a number of proven axioms and lemmas about Verus' seq, set, and map types (as well as a few for multisets). This will make it easier for Verus users to write proofs that use these types.

# Ported Dafny Lemmas

## Sequences

The following table demonstrates the sequence lemmas found in the Dafny library and the corresponding Verus equivalent. The verus equivalent will be implemented by this PR unless otherwise clarified to be already implemented.

| Dafny                  |    Verus         |
| ---------------------- | -----------------|
| LemmaLast              | lemma_add_last_back |
| LemmaAppendLast     |  lemma_append_last   |
| LemmaConcatIsAssociative | lemma_concat_associative |
| IndexingImpliesMembership | lemma_indexing_implies_membership |
| MembershipImpliesIndexing | lemma_membership_implies_indexing |
| LemmaSplitAt | lemma_split_at |
| LemmaElementFromSlice | lemma_element_from_slice |
| LemmaSliceOfSlice  | lemma_slice_of_slice |
| LemmaCardinalityOfSet | lemma_cardinality_of_set |
| LemmaCardinatliyOfEmptySetIs0 | lemma_cardinality_of_empty_set_is_0 |
| LemmaNoDuplicatesInConcat | lemma_no_dup_in_concat |
| LemmaCardinalityOfSetNoDuplicates | unique_seq_to_set (already implemented) |
| LemmaNoDuplicatesCardinalityOfSet | lemma_no_dup_set_cardinality |
| LemmaMultisetHasNoDuplicates | lemma_multiset_has_no_duplicates |
| LemmaZipOfUnzip | lemma_zip_of_unzip |
| LemmaMaxOfConcat | lemma_max_of_concat |
| LemmaMinOfConcat | lemma_min_of_concat |
| LemmaSubseqMax | lemma_subseq_max |
| LemmaSubseqMin | lemma_subseq_min |
| LemmaFlattenConcat | lemma_flatten_concat |
| LemmaFlattenReverseConcat | lemma_flatten_reverse_concat |
| LemmaFlattenAndFlattenReverseAreEquivalent | lemma_flatten_and_flatten_reverse_are_equivalent |
| LemmaFlattenLengthGeSingleElementLength | lemma_flatten_length_ge_single_element_length | LemmaFlattenLengthLeMul | lemma_flatten_length_le_mul |
| LemmaMapDistributesOverConcat | ignored |
| LemmaFilterDistributesOverConcat | ignored |
| LemmaFoldLeftDistributesOverConcat | ignored |
| LemmaInvFoldLeft | ignored |
| LemmaFoldRightDistributesOverConcat | ignored |
| LemmaInvFoldRight | ignored |
| SortedUnique | lemma_sorted_unique |

The ignored lemmas deal with higher-order sequence functions, which I was advised to ignore.

## Sets

| Dafny                  |    Verus         |
| ---------------------- | -----------------|
| LemmaSubset   | definition of subset_of (already implemented) |
| LemmaSubsetSize | lemma_len_subset (already implemented) |
| LemmaSubsetEquality | lemma_subset_equality |
| LemmaSingletonSize | ignored  |
| LemmaSingletonEquality | ignored |
| LemmaIsSingleton | ignored |
| LemmaMapSize | lemma_map_size |
| LemmaFilterSize | lemma_len_filter (already implemented) |
| LemmaUnionSize | lemma_len_union_ind |
| LemmaBoundedSetSize | lemma_int_range |
| LemmaGreatestImpliesMaximal | lemma_greatest_implies_maximal |
| LemmaLeastImpliesMinimal | lemma_least_implies_minimal |
| LemmaMaximalEquivalentGreatest | lemma_maximal_equivalent_greatest |
| LemmaMinimalEquivalentLeast | lemma_minimal_equivalent_least |
|  LemmaLeastIsUnique | lemma_least_is_unique |
|  LemmaGreatestIsUnique | lemma_greatest_is_unique |
| LemmaMinimalIsUnique | lemma_minimal_is_unique |
| LemmaMaximalIsUnique | lemma_maximal_is_unique |
| LemmaFindUniqueMinimal | spec fn find_unique_minimal and proof fn _find_unique_minimal_ensures |
| LemmaFindUniqueMaximal | spec fn find_unique_maximal and proof fn _find_unique_maximal_ensures |

## Maps

| Dafny                  |    Verus         |
| ---------------------- | -----------------|
| LemmaDisjointUnionSize | lemma_disjoint_union_size |
| LemmaInvertIsInjective | lemma_invert_is_injective |

# Ported Dafny Axioms

The Dafny prelude includes a vast number of axioms that relieve some responsibility from the user attempting to get proofs to pass. This PR implements these axioms in the form of proven lemmas. Because they are no longer axiomatized, the facts they provide are not automatically included in the context. The solution to this is a "magic" lemma that, when included in a proof body, brings in the same context as the Dafny prelude. These magic lemmas are so named:
- lemma_seq_properties
- lemma_set_properties
- lemma_map_properties
- lemma_multiset_properties
  
Lemmas brought in from the Dafny prelude axioms are currently labeled `This verified lemma used to be an axiom in the Dafny prelude`.

# Ported Dafny functions

This PR also implements a few more functions for `seq`, `set`, `map`, and `multiset` so that the user has a broader range of options and functionality when dealing with these built-in datatypes.

- Seq:
  - fill (constructor)
  - take
  - drop
  - first
  - is_prefix_of
  - is_suffix_of
  - sort_by for generic types and sort for ints
  - max_via for generic types and max for ints
  - min_via for generic types and min for ints
  - index_of_option
  - first_index_of and first_index_of_option
  - last_index_of and last_index_of_option
  - drop_first
  - to_multiset
  - multiplicity
  - remove_value
  - reverse
  - zip_with
  - unzip
  - flatten and flatten_reverse
- Set
  - to_seq and to_sorted_seq
  - is_singleton
  - find_unique_minimal and find_unique_maximal
  - to_multiset
- Map
  - values
  - len
  - is_equal_on_key
  - is_injective
  - invert
  - is_monotonic and is_monotonic_from
- Multiset
  - new (constructor)
  - update
  - intersection_with
  - difference_with
  - disjoint_with
  - dom

### CODE REVIEW

**Review by tjhance — CHANGES_REQUESTED** [review]
I'm a little worried about adding so many broadcast_forall axioms into code that everybody will be importing. I know Dafny had some trouble with too many sequence-related triggers, causing some really slow verification times whenever you had too many sequence subranges.

In particular, I think `lemma_element_from_slice` looks especially worrying, since it will add `new[pos-a]` into the context when `new[pos]` matches.

And then there are a handful of others, like `lemma_split_at` or `lemma_slice_of_slice` that probably aren't that useful anyway because they prove equalities that would be easy for Verus anyway as long as the user is using `=~=`, which is pretty reliable.

For now, I think we should probably remove the `broadcast_forall` attribute, and make them  normal lemmas with proof bodies that are checked by Verus.

**Comment by tjhance** [comment]
@Chris-Hawblitzel do you have a second opinion on what makes for good broadcast_forall axioms? My feeling is that the minimal-ish set we have right now is close to what we want, at least as far as the basic operations go (like `add` and `push` and whatnot).

**Comment by Chris-Hawblitzel** [comment]
I think we should avoid adding broadcast_forall universally to basic operations like `append`, because lemmas like `lemma_concat_associative` can easily create a lot of quantifier instantiations.  Instead, in the long run, we should be able selectively import sets of lemmas using the proposed https://github.com/verus-lang/verus/discussions/37 .  So for now the safe thing to do is to omit the broadcast_foralls and wait for https://github.com/verus-lang/verus/discussions/37 .

File: source/pervasive/seq_lib.rs
Comment (elaustell) [reply source/pervasive/seq_lib.rs:550]: I'm not so sure about that. Consider the sequence `[[1, 2, 3], [4, 5, 6], [7, 8, 9]]`. Flatten_reverse should result in the sequence `[7, 8, 9, 4, 5, 6, 1, 2, 3]`. Flatten and then reverse would result in `[9, 8, 7, 6, 5, 4, 3, 2, 1]`. Or I might just not be understanding your suggestion...

**Comment by utaal** [comment]
Thank you for porting these over! I have a small suggestion: can we tag the functions and lemmas that you've imported from dafny with a comment? (I'm kind of tempted to maybe even add a marker attribute, for this purpose) This would help us distinguish between the ones that had been added because someone needed them and those that we're inheriting. This can, for example, be helpful in deciding what goes into the import sets when we implement https://github.com/verus-lang/verus/discussions/37 in August/September.

**Comment by jaybosamiya** [comment]
Just making a quick note here that there is a lot of added context to this code that happens in IRL discussions that hasn't made its way to GitHub just yet (esp since this is still a draft, not yet ready for review). Once it is ready for review, we'll explicitly mark it as such and add more reviewers.

Fwiw, multiple things added to this branch at the current moment are not going to directly make it to the final PR (such as many(/most?) of the added `broadcast_forall`s), since some of the code is being used to test things out (a direct port of the Dafny std library wouldn't quite work, due to _reasons_, thus a bunch of testing is needed to see what might work well).

The aim overall though is that things will be prepared well enough for #37.  The way we've been proceeding for that is somewhat different from adding a marker attribute or comment (it'll probably make more sense when the PR is marked ready-to-review; we're still testing more things out).  But yeah, good reminder to track things (our current tracking strategy, at least until we get closer to final PR is just that _everything_ in this branch is tainted, so no need to add comments/attributes 😄)

File: source/pervasive/set.rs
Comment (tjhance) [inline source/pervasive/set.rs:3]: can you remove these notes

File: source/pervasive/set.rs
Comment (jaybosamiya) [inline source/pervasive/set.rs:199]: Isn't this equivalent to `axiom_set_insert_same` + `axiom_set_insert_same`. I don't think this should be a separate axiom.

File: source/pervasive/set_lib.rs
Comment (jaybosamiya) [inline source/pervasive/set_lib.rs:68]: Why is this commented out?

File: source/pervasive/std_specs/vec.rs
Comment (jaybosamiya) [inline source/pervasive/std_specs/vec.rs:248]: Why is this commented out?

**Review by jaybosamiya — COMMENTED** [review]
Thanks for all the work on this! This definitely will make using Verus a lot easier for projects by having a well rounded standard library.

I've added comments inline (use the "Files Changed" view to easily see them all), but also have some open questions that I am not sure I've actually made up my mind about:
1. Do we want lemmas to be methods? I think it might actually make usage nicer (eg: rather than  `lemma_remove_key_len(m, k)`, `m.lemma_remove_key_len(k)` would require you to import fewer things via `use` declarations, and _as a personal preference_ seems to also feel like a cleaner style). Not sure how strong others opinions on this are, so I'm opening this up as a question here.
2. Do we even need to specify the "Ported from Dafny prelude" comment?  Since there are now the `lemma_X_properties` functions that are equivalent to those, I don't think we need to mark the specific individual ones.

Also, I am yet to do a full detailed pass over every single function and comment, but hopefully the current set of comments already point towards the style in which I am looking over things.  I'll do a full pass later on after the next set of changes are pushed.

File: source/pervasive/map.rs
Comment (tjhance) [reply source/pervasive/map.rs:356]: If a spec fn can be defined with a body, it probably makes sense to put it in the `lib` file. Likewise if a lemma has a body, it makes sense to put it in `lib`. The `set.rs` and `seq.rs` files are mostly for spec fns with no bodies, and proof fns marked external_body.

File: source/pervasive/seq_lib.rs
Comment (elaustell) [reply source/pervasive/seq_lib.rs:276]: for both the regular one and the option one?

File: source/pervasive/std_specs/vec.rs
Comment (elaustell) [reply source/pervasive/std_specs/vec.rs:248]: Oh I forgot about this cause I did it very early on, I think I got stuck and confused on it. Anyways I'm just going to delete it because it's not really relevant to this project anyways

**Review by tjhance — APPROVED** [review]

File: source/pervasive/seq_lib.rs
Comment (parno) [inline source/pervasive/seq_lib.rs:901]: There's an extra tab here

File: source/pervasive/seq_lib.rs
Comment (parno) [inline source/pervasive/seq_lib.rs:983]: "0<=" => "0 <="

File: source/pervasive/seq_lib.rs
Comment (parno) [inline source/pervasive/seq_lib.rs:1004]: "minimum"

File: source/pervasive/seq_lib.rs
Comment (parno) [inline source/pervasive/seq_lib.rs:1082]: Probably remove `pub` and use `//`

File: source/pervasive/seq_lib.rs
Comment (parno) [inline source/pervasive/seq_lib.rs:1100]: In Dafny, I suspect this function was executable; here it's a spec.  In theory, for a spec sort, we can use something less efficient than merge sort.  I think we can leave it for now though (maybe it'll make someone's job of writing an exec version easier).

File: source/pervasive/seq_lib.rs
Comment (parno) [inline source/pervasive/seq_lib.rs:1258]: "only all" => "only the"?

File: source/pervasive/seq_lib.rs
Comment (parno) [inline source/pervasive/seq_lib.rs:1297]: No trigger

File: source/pervasive/seq_lib.rs
Comment (parno) [inline source/pervasive/seq_lib.rs:1303]: No trigger

File: source/pervasive/seq_lib.rs
Comment (parno) [inline source/pervasive/seq_lib.rs:1383]: comma between "duplicated" and "and

File: source/pervasive/seq_lib.rs
Comment (parno) [inline source/pervasive/seq_lib.rs:1462]: Drop "Proves that"

File: source/pervasive/seq_lib.rs
Comment (parno) [inline source/pervasive/seq_lib.rs:1546]: No triggers

File: source/pervasive/seq_lib.rs
Comment (parno) [inline source/pervasive/seq_lib.rs:1575]: Many of these still have extraneous triggers

File: source/pervasive/set_lib.rs
Comment (parno) [inline source/pervasive/set_lib.rs:83]: Probably should use `//` and no `pub`

Comment (parno) [reply source/pervasive/set_lib.rs:83]: Similar comment for other helper-style functions below.

File: source/pervasive/set_lib.rs
Comment (parno) [inline source/pervasive/set_lib.rs:275]: if-else formatting

File: source/pervasive/set_lib.rs
Comment (parno) [inline source/pervasive/set_lib.rs:591]: "set union" -> "union"

File: source/pervasive/set_lib.rs
Comment (parno) [inline source/pervasive/set_lib.rs:665]: Some extra trigger annotations in the proofs below

File: source/pervasive/set_lib.rs
Comment (parno) [inline source/pervasive/set_lib.rs:675]: Not sure what this comment means

File: source/rust_verify/example/multiset.rs
Comment (parno) [inline source/rust_verify/example/multiset.rs:26]: "eg" -> "leq"?

File: source/pervasive/multiset.rs
Comment (elaustell) [reply source/pervasive/multiset.rs:563]: Sure, my only thought is it's so ugly that it would be best to hide it at the bottom. Like here are all these lemmas and down here is them tied in a little bow. But I'm happy to move it up if you think that fits better with convention.

File: source/pervasive/seq_lib.rs
Comment (elaustell) [reply source/pervasive/seq_lib.rs:317]: I get the error "in pub open spec function, cannot refer to private function"

File: source/pervasive/seq_lib.rs
Comment (elaustell) [reply source/pervasive/seq_lib.rs:1575]: Yes sorry I forgot to take all the triggers off after removing broadcast formal

File: source/rust_verify/example/multiset.rs
Comment (elaustell) [reply source/rust_verify/example/multiset.rs:26]: I think the point was it's an example of using the function sorted_by. But I can change it if you think that is more appropriate

File: source/pervasive/multiset.rs
Comment (elaustell) [reply source/pervasive/multiset.rs:563]: My only thought is that these make more sense being called from the math library, but when I try to call that verus complains that it's not supported. But these functions don't make a whole lot of sense in the Multiset library, which is why I hid them at the bottom. Thoughts on this?

File: source/rust_verify/example/multiset.rs
Comment (parno) [reply source/rust_verify/example/multiset.rs:26]: Yes, I think the `eg` looks odd without the periods and in a non-formal writing setting.

File: source/pervasive/multiset.rs
Comment (parno) [reply source/pervasive/multiset.rs:564]: Hmm, what error do you see when you put the definitions in the math library?  I think that should be feasible to do.

**Comment by utaal** [comment]
Thank you for all your work on this. Having an extensive standard (spec) library is going to significantly improve the experience of using Verus.

I've left a few comments inline, and I have two general questions:
* there are a number of lemmas that have no proof: did they also have no proof in Dafny?
* there are a number of `open spec` functions with an accompanying `*_ensures` lemma which appear to carry the properties of the function that are useful to the user: were they also not opaque in Dafny? Should we consider marking (some of them) closed? (@jonhnet is working on adding ensures to spec functions, which would make these `_ensures` lemmas more directly usable without needing to call them).

File: source/pervasive/seq_lib.rs
Comment (jaybosamiya) [reply source/pervasive/seq_lib.rs:274]: This one in particular, I had explicitly suggested _against_ `max_by`, and specifically had suggested `max_via`, see https://github.com/verus-lang/verus/pull/622#discussion_r1275397096 but relevant part of the comment:
> leaves room in the future for a `max_by` which can use a convention similar to that in Rust's standard library (see [`max_by`](https://doc.rust-lang.org/std/cmp/fn.max_by.html)) that would use a `FnSpec(A,A)->Ordering`)

File: source/pervasive/seq.rs
Comment (elaustell) [reply source/pervasive/seq.rs:195]: I have no idea why this is there, I wasn't the one who put it there. I'm happy to leave it or remove it.

**Review by Chris-Hawblitzel — APPROVED** [review]

**Review by utaal — APPROVED** [review]
Thank you again for all the work!

(At the meeting, we discussed moving some of the empty lemmas derived from Dafny axioms to tests; we can do this either now or after this PR is merged.)

File: source/pervasive/seq_lib.rs
Comment (utaal) [reply source/pervasive/seq_lib.rs:274]: Sgtm. Thanks for the clarification, and sorry for missing your previous comment.

File: source/pervasive/seq.rs
Comment (utaal) [reply source/pervasive/seq.rs:195]: Please remove it -- it currently has no effect, thanks!

**Comment by utaal** [comment]
Please clean up the history before merging. (Let me know if I can help.)

**Pull request merged**
